cpp/oneapi/dal/detail/memory_impl_host.cpp

```diff
@@ -0,0 +1,73 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/detail/memory_impl_host.hpp"
+
+#include <daal/include/services/daal_memory.h>
+
+namespace oneapi::dal::detail::v1 {
+
+void* malloc_impl_host(const default_host_policy&, std::int64_t size) {
+    auto ptr = daal::services::daal_malloc(detail::integral_cast<std::size_t>(size));
+    if (ptr == nullptr) {
+        throw dal::host_bad_alloc();
+    }
+    return ptr;
+}
+
+void free_impl_host(const default_host_policy&, void* pointer) {
+    daal::services::daal_free(pointer);
```

Comment:
Not sure if `daal_free` allows safe `nullptr` delete, please double check

Reply:
Yes, it works safely: https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/support-functions/memory-management/mkl-free.html